### PR TITLE
dotcom: Don't depend on gitserver for repo.add

### PR DESCRIPTION
**Pull request by eseliger**
The IsRepoCloneable had two hats before:
- Check if the given repo that Sourcegraph knows about is cloneable; this is used for both dotcom and on-prem
- Check if an unknown repo name might be accessible to sourcegraph without any auth provided; this is solely used on dotcom to check if we should lazily add a new repo when visited

This code was confusing to me and nothing prevented it from technically running for on-prem as well, when a repo is lost in between a request for isCloneable coming in and it hitting gitserver - we then might just talk to a random upstream with no auth.

Up until recently, we did that for all repos because we were missing the errcode.IsNotFound check - meaning during DB outages we might have hit external hosts unexpectedly as well.

This also wouldn't have worked for non-git repos, because we just assumed all repos we see are git repos. For example, if we ever added perforce here this wouldn't work.

This PR makes the isRepoCloneable endpoints sole purpose to check if an existing repo can be cloned, using it's actual VCSSyncer implementation.

Instead of reaching out to gitserver, frontend simply checks if the git API endpoint is reachable, not requiring another roundtrip to an internal service, and making it work while gitserver is rolling out as well, as a nice side-effect.

## Test plan

I tested this new logic with both GitHub.com and GitLab.com and it seems to work reliably.

### CODE REVIEW

File: cmd/frontend/backend/repos.go [hunk 1]
```diff
@@ -159,6 +147,69 @@ func (s *repos) addRepoToSourcegraphDotCom(ctx context.Context, name api.RepoNam
...
 
+var metricIsRepoCloneable = promauto.NewCounterVec(prometheus.CounterOpts{
+	Name: "src_frontend_repo_add_is_cloneable",
+	Help: "temporary metric to measure if this codepath is valuable on sourcegraph.com",
+}, []string{"status"})
+
+// isGitRepoPubliclyCloneable checks if a git repo with the given name would be
+// cloneable without auth - ie. if sourcegraph.com could clone it with a cloud_default
+// external service. This is explicitly without any auth, so we don't consume
+// any API rate limit, since many users visit private or bogus repos.
```
Comment (keegancsmith) [inline cmd/frontend/backend/repos.go:158]: depending on how often we hit this, the IP will be taken into account for rate limiting.

I feel like back in the day we had some debouncing per repo name on this check. It isn't clear to me we will still have that.

File: cmd/frontend/backend/repos.go [hunk 2]
```diff
@@ -159,6 +147,69 @@ func (s *repos) addRepoToSourcegraphDotCom(ctx context.Context, name api.RepoNam
...
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	status := "unknown"
+	defer func() {
+		metricIsRepoCloneable.WithLabelValues(status).Inc()
+	}()
+
+	// Speak git smart protocol to check if repo exists without cloning.
+	remoteURL, err := vcs.ParseURL("https://" + string(name) + ".git/info/refs?service=git-upload-pack")
```
Comment (keegancsmith) [inline cmd/frontend/backend/repos.go:174]: should work for the major code hosts, no idea if historically we would work with others. If not, I doubt that was used much?

File: cmd/frontend/backend/repos.go [hunk 3]
```diff
@@ -159,6 +147,69 @@ func (s *repos) addRepoToSourcegraphDotCom(ctx context.Context, name api.RepoNam
...
+
+	resp, err := cf.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "failed to check if repository exists")
+	}
+
+	// No interest in the response body.
+	_ = resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
```
Comment (keegancsmith) [inline cmd/frontend/backend/repos.go:198]: maybe also check if a reasonable http header is present just to make sure this isn't a random website which always returns 200.

I suppose I would feel more comfortable adding an entry to the repo table if indeed it is responding with legit git responses.

**Review by keegancsmith — APPROVED** [review]
Reading the isRepoCloneable handler I am unsure if we did any debouncing, but I feel like this is something we did back in the day since we often had people browsing private repos with the sourcegraph extension.

What does the current metric look like, how often are we hitting this code path?

Either way, this code LGTM from an understanding perspective. Just check out inline comments.

File: cmd/frontend/backend/repos.go (hunk 2 above)
Comment (pjlast) [reply cmd/frontend/backend/repos.go:174]: I don't think the other code hosts would be "publicly cloneable" anyways? 🤔

File: cmd/frontend/backend/repos.go (hunk 3 above)
Comment (pjlast) [reply cmd/frontend/backend/repos.go:198]: This is a good point. Perhaps "adding publicly available git repos" should only be possible from a curated list of sites?

File: cmd/frontend/backend/repos.go [hunk 4]
```diff
@@ -159,6 +147,69 @@ func (s *repos) addRepoToSourcegraphDotCom(ctx context.Context, name api.RepoNam
...
+		if ctx.Err() != nil && err == ctx.Err() {
```
Comment (ggilmore) [inline cmd/frontend/backend/repos.go:199]:
```suggestion
		if err := ctx.Err() != nil && err != nil {
```

File: cmd/frontend/backend/repos.go (hunk 1 above)
Comment (eseliger) [reply cmd/frontend/backend/repos.go:158]: The gitserver implementation doesn't have it, and we don't seem to have run into trouble in the past months with it 🤔 
I agree some caching here could be useful though, maybe a simple redis KV with 1h expiry or so. 
Will need some testing to find good numbers, saving for a later PR :)

File: cmd/frontend/backend/repos.go (hunk 2 above)
Comment (eseliger) [reply cmd/frontend/backend/repos.go:174]: the check above for if it's from a public code host will only let github and gitlab dotcom get here, so I think this is ok. These two work with this URL scheme (tested by hand)

File: cmd/frontend/backend/repos.go (hunk 3 above)
Comment (eseliger) [reply cmd/frontend/backend/repos.go:198]: This is just a preflight check - repo updater will do a proper authed source.GetRepo(ctx, name) call to verify again. This is just to save our token from being used too much, at least that's what I deduced from the existing code.

File: cmd/frontend/backend/repos.go (hunk 4 above)
Comment (eseliger) [reply cmd/frontend/backend/repos.go]: ah, ty!

**Review by ggilmore — APPROVED** [review]

**Pull request merged**